README.md
[![Python application](https://github.com/alexhallam/tablespoon/actions/workflows/python-app.yml/badge.svg)](https://github.com/alexhallam/tablespoon/actions/workflows/python-app.yml)

<h1 align="center">tablespoon</h1>
<p align="center"><b>T</b>ime-series <b>B</b>enchmark methods that are <b>S</b>imple and <b>P</b>robabilistic</p>

<p align="center"><img align="center" src="assets/tbsp.png" width="100" /></p>


# Documentation and quick links
- [Documentation and quick links](#documentation-and-quick-links)
- [Introduction](#introduction)
- [Why Run Simple Methods](#why-run-simple-methods)
- [Quick Example](#quick-example)
    - [Seasonal Example](#seasonal-example)
    - [Stock Prediction](#stock-prediction)
- [Goals of this package](#goals-of-this-package)
- [Non-Goals](#non-goals)
- [Installation](#installation)
    - [Python](#python)
- [Citation](#citation)
- [References](#references)
- [Recommended probabilistic forecasting packages](#recommended-probabilistic-forecasting-packages)
- [Learn more about forecasting](#learn-more-about-forecasting)
- [Built with poetry and pushed to pypi](#built-with-poetry-and-pushed-to-pypi)

# Introduction

Many methods exist for probabilistic forecasting. If you are looking for an
impressive probabilistic forecasting package see the list of recommendation at
the bottom of this README. This package is <b>exceptionally ordinary</b>. It is
expected that this package may be used as a compliment to what is already out
there.

# Why Run Simple Methods

We have found, by experience, many good uses for the methods in this package.
To often we see that forecast methods go in production without a naive method to
accompany it. This is a missed opportunity.

1. **Naive May Be Good Enough**: How good is good enough? I have almost never started a project where the firm receiving the forecast knew the answer to this question, but it is important to get to it early. I think all of us would agree that a forecast error of 0 is best, but in reality how much time should be spent spinning our wheels to get there. Further, many business applications can’t be improved upon even with significant forecast error reduction. I remember hearing a story of a firm which ordered units on pallets. The firm paid by the pallet, not the unit. The forecast team was able to reduce the forecast error, but in the end this did not change the outcome. This is the ‘pallet problem’. As data scientists it is easy to think of error reduction as a continuous function, but this need not map the same way to business value. In this example the continuous error reduction mapped to a discrete pallet. This is a long way of saying, “A naive forecast may be good enough”. What is a naive forecasting method? I typically think of two methods. The mean forecast and the naive forecast. I will go over these in more detail below. There are also some applications where naive methods have been shown to be hard to beat.
2. **Get A Denominator for Relative Metrics**: Though naive methods can usually
   be beat it is good to know the relative improvement over the benchmark. This
   can allow a forecasting team to market their alternative forecast when the
   'skill score' is impressive.
3. **Easy to productionize and get expectations**: Get a sense for *how good is
   good enough*. In many applications a forecast team is asked to forecast, but
   stakeholders provide no line-in-the-sand for when the forecasting work needs
   to stop. One reasonable approach is to run the benchmarks found in this
   package then beat the best performing benchmark by a margin that is
   statistically significant.
4. **Resilience in Production - Why not have many models?**: Sometimes, despite
   best efforts the production model does something unexpected. In this
   case it is nice to have a simple backup that is cheap to generate and good
   enough to fall back on. In this way a production forecast pipeline gains
   strength from a diversity of models.
5. **Easy Uncertainty Quantification**: Recently we see that applications
   are not about forecast accuracy, but instead about forecasting uncertainty.
   Capturing the full distribution helps firms set "service levels" aka
   percentiles of the distribution for which they are prepared to serve. Even
   if you have the worlds most accurate unbiased forecast the median point is
   , by definition, an underforecast half the time. For this reason it is best to provide a
   distribution of simulated future values and the firm may decide for
   themselves what risks they are or are not willing to take.

# Quick Example

We show a quick example below. 

For more examples see [Simple Example](https://alexhallam.github.io/tablespoon/section/plotting/), [Extended Example](https://alexhallam.github.io/tablespoon/section/extended/)

There are also some Notebook examples
1. [Intro](https://github.com/alexhallam/tablespoon/blob/main/examples/tablespoon.ipynb)
2. [Time Series Cross Validation & Skill Score](https://github.com/alexhallam/tablespoon/blob/main/examples/time_series_cv.ipynb)
### Seasonal Example

```python
import tablespoon as tbsp
from tablespoon.data import SEAS

sn = tbsp.Snaive()
df_sn = sn.predict(
    SEAS, horizon=7 * 4, frequency="D", lag=7, uncertainty_samples=8000
).assign(model="snaive")

print(df_sn.head(10))
```

### Stock Prediction

```python
import tablespoon as tbsp
from tablespoon.data import APPL

n = tbsp.Naive()
df_n = n.predict(
    APPL, horizon=7 * 4, frequency="D", lag=1, uncertainty_samples=8000
).assign(model="naive")

print(df_n.head(10))
```

```sh
          ds  rep    y_sim  model
0 2022-01-02    0  5.20006  naive
1 2022-01-02    1  5.16789  naive
2 2022-01-02    2  5.17641  naive
3 2022-01-02    3  5.19340  naive
4 2022-01-02    4  5.20075  naive
5 2022-01-02    5  5.17681  naive
6 2022-01-02    6  5.20302  naive
7 2022-01-02    7  5.18896  naive
8 2022-01-02    8  5.19622  naive
9 2022-01-02    9  5.17469  naive
```
<p align="center"><img align="center" src="assets/forecasts_n.jpg" width="800" /></p>




# Goals of this package

1. ♙**Simple**: Not just in the forecasts themselves, but also from the
   users perspective.
2. ♝**Documented**: It should be very clear exactly how forecasts are getting
   generated. We document the parameterization of the models to make this as
   obvious and uninteresting as possible. See [Forecast Method Math Documentation](https://alexhallam.github.io/tablespoon/section/math/)
3. ♜**Stable**: We want this package to feel rock solid. For this to happen
   we keep the feature set relatively small. We believe that after the initial 
   development of this package we should spend out time maintaining the code as
   oppose to thinking of new features.
4. ♞**Distributional**: Quantification of uncertainty is the name of
   the game.

# Non-Goals

1. 🔥**Circut Melting Focus on Speed**: Not to say this is a slow package. In fact, it is very fast! 
   We just don't put any extra effort to make it faster than `numpy`.
2. 🤖**New/Complex Forecast Models**: Again, this is out of scope. If you are
   looking for recommendations please see the bottom of the page.

# Installation

### Python

```
pip3 install tablespoon
```

# Citation

If you would like to cite `tablespoon`, please cite it as follows:

Alex Hallam. **tablespoon: Time-series Benchmark methods that are Simple and Probabilistic** https://github.com/alexhallam/tablespoon, 2022. Version 0.4.6.

```
@misc{tablespoon,
  author={Alex Hallam},
  title={{tablespoon}: {Time-series Benchmark methods that are Simple and Probabilistic},
  howpublished={https://github.com/alexhallam/tablespoon},
  note={Version 0.4.6,
  year={2022}
}
```

# References

1. Hyndman, R.J., & Athanasopoulos, G. (2021) Forecasting: principles and practice, 3rd edition, OTexts: Melbourne, Australia. OTexts.com/fpp3. Accessed on 2021-09-26.

# Recommended probabilistic forecasting packages

There are many packages that can compliment `tablespoon`

[forecast](https://github.com/robjhyndman/forecast): The king of forecasting
packages. Rob Hyndman is a professor of forecasting and has served as editor of
the journal "International Journal of Forecasting". If you are new to
forecasting please read his free ebook [fpp3](https://otexts.com/fpp3/).

[prophet](https://facebook.github.io/prophet/): A very capable and reliable
forecasting package. I have never seen a bad forecast come out of prophet.

[gluonts](https://ts.gluon.ai/). If you are itching to use neural nets for
forecasting this is a good one to pick.

# Learn more about forecasting

1. Read [fpp3](https://otexts.com/fpp3/)
2. Join the [International Institute of Forecasting](https://forecasters.org/)
   and read their publications.
   
# Built with poetry and pushed to pypi

```sh
poetry publish -u <username> -p <password> --build
```


CHANGELOG.md
# 0.4.6 Adding TimeSeriesInitialSplit

In this update I added `TimeSeriesInitialSplit`. This will split the data in a convenient way for time series testing.

# 0.4.5 Droping Stan

I love Stan, but I had to drop it as the backend sampler.

A lot of problems were solved as this package was developed. Multiprocessing
with cmdstanpy folder structures was a big one. As I used this package more
in production I started to dread it more than love it. Why? A lot of small
simple things. Keeping tabs on `cmdstanpy`, getting a version of python
that all of the dependancies were happy with felt tedious. In prod environments
you also have to figure out if you have C++ compilation tools like gcc. 
This is not always the case. If those tools are not available then `install_cmdstanpy`
will not work. This is additional overhead, and something I did not want to
have to worry about. Once I got
a python version 3.8>= to work well, I found that when I wanted to make 
a notebook example with Google Colab I had to wrestle with the notebook
to get a more recent version of python -- Colab comes with 3.7 by default.
The version was not all though. I had to make sure that the priority was 
reset to run the more recent version. This was not the end. I also had 
to install additional software that comes with the newer version of 
python, but does not come with the notebook -- something called `dsutils`?

Why would I care so much about the notebook environment? Well, I think 
there are a population of people who like to run code on notebooks. This
could be how they 'kick the tires' when trying a new package. If that 
experience goes well then maybe the chances of making into production is 
higher.

Though I love Stan, and it hurts to remove it, I think that these models
are simple enough to survive just fine without stans sampler. 

If you are interested in how to use stan with a python then checkout out 
version *4.4* of this package commit hash `49066323c08915e90b8b1e06a88cbddb52a726ae`

I think these changes will make the package easier for me to maintain and
very simple to integrate for production runs.


tests/test_tablespoon.py
from tablespoon import __version__

# For testing locally
# poetry run python tests/test_tablespoon.py

# For dev run
# poetry run pytest

import numpy as np
import pandas as pd
from tablespoon import Naive, Mean, Snaive
from tablespoon.data import WALMART_TX, SEAS
from tablespoon.model_selection import TimeSeriesInitialSplit
import pytest


# small snaive test
df = SEAS
n = Snaive()
df_n = n.predict(
    df,
    horizon=7,
    frequency="D",
    lag=7,
    uncertainty_samples=1000,
).assign(model="snaive").loc[:, ['ds', 'y_sim']].groupby('ds', as_index=False).mean()

# the following checks that means are withing 5 percent (0.05) of the last week


def test_snaive_1():
    value = df_n.y_sim[0]
    assert value == pytest.approx(150, 0.05)


def test_snaive_2():
    value = df_n.y_sim[1]
    assert value == pytest.approx(250, 0.05)


def test_snaive_3():
    value = df_n.y_sim[2]
    assert value == pytest.approx(350, 0.05)


def test_snaive_4():
    value = df_n.y_sim[3]
    assert value == pytest.approx(450, 0.05)


def test_snaive_5():
    value = df_n.y_sim[4]
    assert value == pytest.approx(550, 0.05)


def test_snaive_6():
    value = df_n.y_sim[5]
    assert value == pytest.approx(650, 0.01)


def test_snaive_7():
    value = df_n.y_sim[6]
    assert value == pytest.approx(750, 0.01)


# larger snaive test
df = WALMART_TX
n = Snaive()
df_tx = n.predict(
    df,
    horizon=14,
    frequency="D",
    lag=7,
    uncertainty_samples=1000,
).assign(model="snaive").loc[:, ['ds', 'y_sim']].groupby('ds', as_index=False).mean()
print(df_tx)


def test_snaive_tx_1():
    value = df_tx.y_sim[0]
    assert value == pytest.approx(12228, 0.20)


def test_snaive_tx_2():
    value = df_tx.y_sim[1]
    assert value == pytest.approx(11370, 0.20)


def test_snaive_tx_3():
    value = df_tx.y_sim[2]
    assert value == pytest.approx(10375, 0.20)


def test_snaive_tx_4():
    value = df_tx.y_sim[3]
    assert value == pytest.approx(9162, 0.20)


def test_snaive_tx_5():
    value = df_tx.y_sim[4]
    assert value == pytest.approx(12303, 0.20)


def test_snaive_tx_6():
    value = df_tx.y_sim[5]
    assert value == pytest.approx(13681, 0.20)


def test_snaive_tx_7():
    value = df_tx.y_sim[6]
    assert value == pytest.approx(14815, 0.20)


# pull and clean data
df = (
    pd.read_csv("https://storage.googleapis.com/data_xvzf/m5_state_sales.csv")
    .query("state_id == 'CA'")
    .rename(columns={"date": "ds", "sales": "y"})
    .assign(y=lambda df: np.log(df.y))
)

# make lag of 1 default
m = Naive()
naive_forecast = m.predict(df, horizon=10, frequency="D")


def test_naive_2016_06_01_lower():
    value = round(
        naive_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.05), 2
    )
    assert value == pytest.approx(7.15, 0.3)


def test_naive_2016_06_01_upper():
    value = round(
        naive_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.95), 2
    )
    assert value == pytest.approx(13.12, 0.3)


m = Mean()
mean_forecast = m.predict(df, horizon=10, frequency="D")
round(mean_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.05), 2)
round(mean_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.95), 2)


def test_mean_2016_06_01_lower():
    value = round(
        mean_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.05), 2
    )
    assert value == pytest.approx(8.85, 0.3)


def test_mean_2016_06_01_upper():
    value = round(
        mean_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.95), 2
    )
    assert value == pytest.approx(10.32, 0.3)


m = Snaive()
snaive_forecast = m.predict(df, horizon=10, frequency="D")
round(snaive_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.05), 2)
round(snaive_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.95), 2)


def test_snaive_2016_06_01_lower():
    value = round(
        snaive_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.05), 2
    )
    assert value == pytest.approx(8.39, 0.3)


def test_snaive_2016_06_01_upper():
    value = round(
        snaive_forecast.query("ds == '2016-06-01'").loc[:, "y_sim"].quantile(q=0.95), 2
    )
    assert value == pytest.approx(10.96, 0.3)


def test_initial_time_series_split():
    X = np.arange(0, 30)
    tscv = TimeSeriesInitialSplit(initial=7 * 3, increment_size=7, gap=0)
    for train_index, test_index in tscv.split(X):
        assert len(train_index) == (7 * 3)
        assert len(test_index) == 7


tablespoon/model_selection.py
from numpy import arange


class TimeSeriesInitialSplit():
    """Time Series cross-validator with initial period

    Provides time series splits for rolling origin type 
    cross validation. This means users may set an initial
    time period. gap size, and increment_size.

    Parameters:
        initial : int, default=21
            Number of splits.
        increment_size : int, default=7
            Sets the size of the test set to be added at each iteration
        gap : int, default=0
            Number of samples to exclude from the end of each train set before
            the test set.
    Examples:
    ```py
    import numpy as np
    from tablespoon.model_selection import TimeSeriesInitialSplit
    X = np.arange(0,50)
    tscv = TimeSeriesInitialSplit()
    for train_index, test_index in tscv.split(X):
        print("TRAIN:", train_index, "TEST:", test_index)
        X_train, X_test = X[train_index], X[test_index]

    > TRAIN: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20] TEST: [21 22 23 24 25 26 27]
    > TRAIN: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27] TEST: [28 29 30 31 32 33 34]
    > TRAIN: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34] TEST: [35 36 37 38 39 40 41]
    > TRAIN: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41] TEST: [42 43 44 45 46 47 48]
    ```
    """

    def __init__(self, initial=7 * 3, increment_size=7, gap=0):
        self.initial = initial
        self.increment_size = increment_size
        self.gap = gap

    def split(self, X):
        """Generate indices to split data into training and test set.
        Parameters:
            X : array-like of shape (n_samples, n_features)
                Training data, where `n_samples` is the number of samples
                and `n_features` is the number of features.
        Yields: 
            train : ndarray
                The training set indices for that split.
            test : ndarray
                The testing set indices for that split.
        """
        n_samples = len(X)
        initial = self.initial
        gap = self.gap
        increment_size = self.increment_size

        # Make sure we have enough samples for the given split parameters
        if initial > n_samples:
            raise ValueError(
                f"Cannot have number of initial_size={initial} greater"
                f" than the number of samples={n_samples}."
            )
        if n_samples - initial - increment_size - gap < 0:
            raise ValueError(
                f"Size of initial + increment_size + gap too large given sample"
                f"={n_samples} with initial={initial} increment_size={increment_size} and gap={gap}."
            )

        indices = arange(n_samples)
        test_starts = range(initial, n_samples, increment_size)
        for test_start in test_starts:
            test = indices[test_start + gap: test_start + increment_size + gap]
            if len(test) < increment_size:
                # break if the test set is smaller than a complete increment_size
                break
            else:
                yield (
                    indices[:test_start],
                    indices[test_start + gap: test_start + increment_size + gap],
                )


tablespoon/forecasters.py
import numpy as np
import pandas as pd
import statsmodels.api as sm
from scipy.stats import norm


def get_sorted_dates(df_historical):
    """
    get historical dates and sort them
    """
    return pd.to_datetime(
        pd.Series(df_historical["ds"].unique(), name="ds")
    ).sort_values()


def check_historical_dates_exist(df_historical):
    """
    check that historical dates exist
    """
    if df_historical["ds"] is None:
        raise Exception("Please include historical dates and name the columns 'ds'.")


def send_helpful_frequency_error():
    """
    check that frequency argument exists else provide helpfull guidance
    """
    raise Exception(
        "Please specify frequency of data. \n`D`- daily\n`1H` - hourly\n For more frequency aliases see https://pandas.pydata.org/docs/user_guide/timeseries.html#timeseries-offset-aliases"
    )


def check_historical_dates_are_contiguous(
    history_dates, min_date, last_date, frequency
):
    """
    check that historical dates are contiguous
    """
    check_dates = pd.date_range(
        start=min_date,
        end=last_date,  # An extra in case we include start
        freq=frequency,
    )  # 'M','D', etc.
    expected_dates = set(check_dates.unique())
    given_dates = set(history_dates.unique())
    remaining_dates = expected_dates.difference(given_dates)
    # print(remaining_dates)
    if len(check_dates) != len(history_dates):
        raise Exception(
            "The series starts on "
            + str(min_date)
            + " and ends on "
            + str(last_date)
            + " the length of the series is expected to be "
            + str(len(check_dates))
            + ", but "
            + str(len(history_dates))
            + " was found "
        )


class Naive(object):
    """Naive Forecaster

    Args:
        object (None): instantiates a Naive Forecast object
    """

    def __init__(
        self,
        history_dates=None,
        include_history=False,
        y=None,
    ):
        self.include_history = include_history
        self.history_dates = history_dates

    def predict(
        self,
        df_historical,
        horizon=30,
        frequency=None,
        lag=1,
        uncertainty_samples=5000,
        include_history=False,
    ):
        """Predict - forecast method

        Args:
            df_historical (pd.DataFrame): A date sorted dataframe with the columns `ds` and `y`
            horizon (int, optional): Forecast horizon. Defaults to 30.
            frequency (int, optional): number of rows that make a seasonal period. Defaults to None.
            lag (int, optional): number of rows that make a seasonal period. Defaults to 1.
            uncertainty_samples (int, optional): number of uncertainty samples to draw. Defaults to 5000.
            include_history (bool, optional): include history. Defaults to False.
            chain_ids (str, optional): identifiers for chain ids. Defaults to None.
            verbose (bool, optional): verbose. Defaults to False.

        Returns:
            pd.DataFrame: A dataframe of predictions as `y_sim`

        Example:
            ```py
            import pandas as pd
            import tablespoon as tbsp
            from tablespoon.data import APPL
            df_APPLE = APPL
            df_APPLE = df_APPLE.assign(ds = lambda df: pd.to_datetime(df.ds))
            naive = tbsp.Naive()
            df_f = (naive.predict(df_APPLE, horizon=7*4, frequency="D", lag = 1, uncertainty_samples = 500).assign(model = 'naive'))
            df_f.head(10)
            ```
        """
        if frequency is None:
            send_helpful_frequency_error()
        self.y = df_historical["y"]
        self.history_dates = get_sorted_dates(df_historical)
        last_date = self.history_dates.max()
        min_date = self.history_dates.min()
        check_historical_dates_are_contiguous(
            self.history_dates, min_date, last_date, frequency
        )
        dates = pd.date_range(
            start=last_date, periods=horizon + 1, freq=frequency
        )  # An extra in case we include start  # 'M','D', etc.
        dates = dates[dates > last_date]  # Drop start if equals last_date
        dates = dates[:horizon]  # Return correct number of periods
        if include_history:
            dates = np.concatenate((np.array(self.history_dates), dates))
        df_dates = pd.DataFrame({"ds": dates})
        df_samples = pd.DataFrame({"rep": np.arange(uncertainty_samples)})
        df_cross = df_dates.merge(df_samples, how="cross")
        # fit
        t = lag + 1
        t_lag = t - lag
        y = self.y.to_numpy()
        end = len(y) - lag
        yt = y[t:]
        yt_lag = y[t_lag:end]
        y_last = y.take(-1)
        mod = sm.GLM(yt, yt_lag, family=sm.families.Gaussian())
        sigma = np.sqrt(mod.fit().scale)
        rng = np.random.default_rng()
        forecast = np.empty([uncertainty_samples, horizon])
        for h in range(0, horizon):
            forecast[:, h] = y_last + sigma * np.sqrt(h + 1) * rng.standard_normal(
                uncertainty_samples
            )
        np_predictions = forecast.transpose().reshape(
            uncertainty_samples * horizon, 1
        )
        df_pred = pd.DataFrame(np_predictions, columns=["y_sim"])
        df_result = pd.concat([df_cross, df_pred], axis=1)
        return df_result


class Mean(object):
    """
    Mean Forecaster
    """

    def __init__(
        self,
        history_dates=None,
        include_history=False,
        y=None,
    ):
        self.include_history = include_history
        self.history_dates = history_dates

    def predict(
        self,
        df_historical,
        horizon=30,
        frequency=None,
        uncertainty_samples=5000,
        include_history=False,
    ):
        """Predict - forecast method

        Args:
            df_historical (pd.DataFrame): A date sorted dataframe with the columns `ds` and `y`
            horizon (int, optional): Forecast horizon. Defaults to 30.
            frequency (int, optional): number of rows that make a seasonal period. Defaults to None.
            lag (int, optional): number of rows that make a seasonal period. Defaults to 1.
            uncertainty_samples (int, optional): number of uncertainty samples to draw. Defaults to 5000.
            include_history (bool, optional): include history. Defaults to False.
            chain_ids (str, optional): identifiers for chain ids. Defaults to None.
            verbose (bool, optional): verbose. Defaults to False.

        Returns:
            pd.DataFrame: A dataframe of predictions as `y_sim`

        Example:
            ```py
            import pandas as pd
            import tablespoon as tbsp
            from tablespoon.data import APPL
            df_APPLE = APPL
            df_APPLE = df_APPLE.assign(ds = lambda df: pd.to_datetime(df.ds))
            mean = tbsp.Mean()
            df_f = (n.predict(df_APPLE, horizon=7*4, frequency="D", lag = 1, uncertainty_samples = 500).assign(model = 'mean'))
            df_f.head(10)
            ```
        """
        self.y = df_historical["y"]
        self.history_dates = get_sorted_dates(df_historical)
        last_date = self.history_dates.max()
        min_date = self.history_dates.min()
        check_historical_dates_are_contiguous(
            self.history_dates, min_date, last_date, frequency
        )
        dates = pd.date_range(
            start=last_date, periods=horizon + 1, freq=frequency
        )  # An extra in case we include start  # 'M','D', etc.
        dates = dates[dates > last_date]  # Drop start if equals last_date
        dates = dates[:horizon]  # Return correct number of periods
        if include_history:
            dates = np.concatenate((np.array(self.history_dates), dates))
        df_dates = pd.DataFrame({"ds": dates})
        df_samples = pd.DataFrame({"rep": np.arange(uncertainty_samples)})
        df_cross = df_dates.merge(df_samples, how="cross")
        # fit
        y = self.y
        T = len(y)
        deg_freedom = T - 1
        mu, sigma = norm.fit(y)
        rng = np.random.default_rng()
        forecast = np.empty([uncertainty_samples, horizon])
        for h in range(0, horizon):
            forecast[:, h] = mu + sigma * np.sqrt(1 + (1 / T)) * rng.standard_t(
                df=deg_freedom, size=uncertainty_samples
            )
        np_predictions = forecast.transpose().reshape(
            uncertainty_samples * horizon, 1
        )
        df_pred = pd.DataFrame(np_predictions, columns=["y_sim"])
        df_result = pd.concat([df_cross, df_pred], axis=1)
        return df_result


class Snaive(object):
    """
    Seasonal Naive Forecaster
    """

    def __init__(
        self,
        history_dates=None,
        include_history=False,
        y=None,
    ):
        self.include_history = include_history
        self.history_dates = history_dates
        self.include_history = include_history

    def predict(
        self,
        df_historical,
        horizon=30,
        frequency=None,
        lag=7,
        uncertainty_samples=5000,
        include_history=False,
    ):
        """Predict - forecast method

        Args:
            df_historical (pd.DataFrame): A date sorted dataframe with the columns `ds` and `y`
            horizon (int, optional): Forecast horizon. Defaults to 30.
            frequency (int, optional): number of rows that make a seasonal period. Defaults to None.
            lag (int, optional): number of rows that make a seasonal period. Defaults to 7 (7 days of a week).
            uncertainty_samples (int, optional): number of uncertainty samples to draw. Defaults to 5000.
            include_history (bool, optional): include history. Defaults to False.
            chain_ids (str, optional): identifiers for chain ids. Defaults to None.
            verbose (bool, optional): verbose. Defaults to False.

        Returns:
            pd.DataFrame: A dataframe of predictions as `y_sim`

        Example:
            ```py
            import tablespoon as tbsp
            from tablespoon.data import SEAS
            sn = tbsp.Snaive()
            df_f = sn.predict(SEAS, horizon=7 * 4, frequency="D", lag=7, uncertainty_samples=800).assign(model="snaive")
            df_f.head(10)
            ```
        """
        self.y = df_historical["y"]
        self.history_dates = get_sorted_dates(df_historical)
        last_date = self.history_dates.max()
        min_date = self.history_dates.min()
        check_historical_dates_are_contiguous(
            self.history_dates, min_date, last_date, frequency
        )
        dates = pd.date_range(
            start=last_date, periods=horizon + 1, freq=frequency
        )  # An extra in case we include start  # 'M','D', etc.
        dates = dates[dates > last_date]  # Drop start if equals last_date
        dates = dates[:horizon]  # Return correct number of periods
        if include_history:
            dates = np.concatenate((np.array(self.history_dates), dates))
        df_dates = pd.DataFrame({"ds": dates})
        df_samples = pd.DataFrame({"rep": np.arange(uncertainty_samples)})
        df_cross = df_dates.merge(df_samples, how="cross")
        # fit
        y = self.y.to_numpy()
        last_start = len(y) - lag
        last_end = len(y)
        yt = y[lag:last_end]
        yt_lag = y[0:last_start]
        mod = sm.GLM(yt, yt_lag, family=sm.families.Gaussian())
        sigma = np.sqrt(mod.fit().scale)
        rng = np.random.default_rng()
        forecast = np.empty([uncertainty_samples, horizon])
        for h in range(0, horizon):
            forecast[:, h] = y[(len(y)) - (lag - ((h) % lag))] + sigma * np.sqrt(np.trunc(((h) * 1) / (lag)) + 1) * rng.standard_normal(uncertainty_samples)
        np_predictions = forecast.transpose().reshape(
            uncertainty_samples * horizon, 1
        )
        df_pred = pd.DataFrame(np_predictions, columns=["y_sim"])
        df_result = pd.concat([df_cross, df_pred], axis=1)
        return df_result
